Epic 8493: WNPRC Purchasing Email Notification

labkey-bpatel opened the pull request:
#### Rationale
Implement email notification upon form submission for a new request submission, request over 5000k, line item changes, request approval or rejection, full quantity received; and send emails to appropriate users.
Spec [here](https://docs.google.com/document/d/1AKhnqOfVJz02K4SR7qY4ohL7jV03gUr-FxAv-dkCIb8/edit#heading=h.ddrj055296t1).

#### Changes
* Create email templates for a new request, line item changes, request status changes.
* Update SubmitRequestAction to send new and updated requests email notification to appropriate users.
* Handle email notif for requests over 5k and less than 5k.
* Handle line item quantity change and full quantity received.
* Handle request status update - mainly Request Rejected, Request Approved, & Order Placed.
* Update data entry form to make it read only for requesters (link in the email will take requesters to this read only data entry page to view the request they submitted).
* Fix issue where Order Complete was not being set correctly.
* Update landing page and admin page url so that they open in a new tab.
* Update default paymentOptions data.
* Update to Receiver's page default view to include the requester.

labkey-bpatel committed:
New request email notification and email template

labkey-bpatel committed:
Changes to ensure new request's email template is being utilized.

labkey-bpatel committed:
Changes to initial impl for request status change and new request.

labkey-bpatel committed:
Line items email notification. Add requester to Receiver default view. Update landing page and admin page url so that they open in a new tab. Update default paymentOptions data.

labkey-bpatel committed:
Fix issue where Order Complete status was not being set.

labkey-bpatel committed:
Handle email notifications for requests over 5000k.

labkey-bpatel committed:
Make request entry form readonly for the requesters/lab end users in order to view the requests created by them.

labkey-bpatel committed:
Update links to avoid js call so that it can be opened on a separate tab. Minor perm change to data entry page.

labkey-bpatel committed:
Handle line item change notification.

labkey-bpatel committed:
Auto test setup update

labkey-bpatel committed:
Line item notif update/refinement. Email receipient related updates. Other minor fixes. Test setup updates.

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/LineItemChangeEmailTemplate.java:
```diff
@@ -0,0 +1,221 @@
+                    ActionURL linkUrl = new ActionURL(WNPRC_PurchasingController.PurchasingRequestAction.class, getContainer());
+                    linkUrl.addParameter("requestRowId", _notificationBean.getRowId());
+                    linkUrl.addParameter("returnUrl", new ActionURL(WNPRC_PurchasingController.RequesterAction.class, c).getPath());
+                    return linkUrl.getBaseServerURI() + c.getPath() + linkUrl.toString();
```
Can you use getURIString() instead?

labkey-bpatel replied:
yes!

labkey-bpatel replied:
well, i'll have to revert this change back, getURIString() doesn't have the container path (i think thats the reason for my initial impl)

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/RequestStatusChangeEmailTemplate.java:
```diff
@@ -0,0 +1,106 @@
+                if (_notificationBean != null)
+                {
+                    if (_notificationBean.getTotalCost().compareTo(BigDecimal.valueOf(5000.0)) >= 0
+                            & (_notificationBean.getRequestStatus().equalsIgnoreCase("Request Approved")
```
Should this be &&?

labkey-bpatel replied:
yes! Good catch!
weird, all those use cases were working as expected in my testing, i must have fat fingered or something before pushing the changes or it was working as-is by coincidence (yikes!)

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -112,19 +152,468 @@ public void addNavTrail(NavTree root) { }
+                columns.add(FieldKey.fromString("qcState/Label"));
+                final Map<FieldKey, ColumnInfo> colMap = QueryService.get().getColumns(tableInfo, columns);
+                TableSelector tableSelector = new TableSelector(tableInfo, colMap.values(), filter, null);
+                oldStatusVal = String.valueOf(tableSelector.getMap().get("qcstate_fs_label"));
```
Don't rely on knowing the generated alias - get the ColumnInfo from the colMap and use its getAlias() method

labkey-bpatel replied:
oh ok, ive seen this in a few places, so thought it was ok. didn't realize getAlias() would do the trick, now changed

labkey-jeckels replied:
I'm less nervous about it when it's a column that's directly on a table instead of a lookup, but still best to let Query choose whatever it wants during SQL generation

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -112,19 +152,468 @@ public void addNavTrail(NavTree root) { }
+                            message.setRecipient(Message.RecipientType.TO, new InternetAddress(user.getEmail()));
+                            MailHelper.send(message, getUser(), getContainer());
+                        }
+                        catch (javax.mail.internet.AddressException | NullPointerException e)
```
When might you get a NullPointerException? Can you add a null check instead?

labkey-bpatel replied:
I think it was from one of my previous impl, it is not needed, now removed

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -112,19 +152,468 @@ public void addNavTrail(NavTree root) { }
+            Map<Integer, String> purchasingDirUserIds = new HashMap<>();
+            for (RoleAssignment roleAssignment : getContainer().getPolicy().getAssignments())
+            {
+                if (roleAssignment.getRole().getName().equals("WNPRC Purchasing Director"))
```
Can you reference a shared constant instead of hard-coding this value here?

labkey-bpatel replied:
done

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -112,19 +152,468 @@ public void addNavTrail(NavTree root) { }
+            String emailSubject = requestEmailTemplate.renderSubject(getContainer());
+            String emailBody = requestEmailTemplate.renderBody(getContainer());
+
+            if (emailTemplateForm.getTotalCost().compareTo(BigDecimal.valueOf(5000.0)) >= 0)
```
Please introduce a constant - there are a couple of places that hard-code the value

labkey-bpatel replied:
done

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -112,19 +152,468 @@ public void addNavTrail(NavTree root) { }
+                            getContainer(), UserManager.getUser(getUser().getUserId()), user,
+                            emailSubject, emailBody,
+                            getDataEntryFormUrl(emailTemplateForm.getRowId(),  new ActionURL(WNPRC_PurchasingController.PurchaseAdminAction.class, getContainer())),
+                            String.valueOf(emailTemplateForm.getRowId()), "New request over 5000k");
```
5000k is $5,000,000!

labkey-bpatel replied:
lol! text is now changed

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -112,19 +152,468 @@ public void addNavTrail(NavTree root) { }
+        User _requester;
+        BigDecimal _totalCost;
+
+        DateFormat _dateFormat = new SimpleDateFormat("yyyy-MM-dd");
```
Is there a good reason to not use the folder's configured date format?

labkey-bpatel replied:
Do you mean from Look & Feel settings?

labkey-jeckels replied:
Yes. In general, we should use that setting unless there's a reason to hard-code something else (like an external system requiring a particular format in a file we're exporting)

labkey-bpatel replied:
done!

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -159,6 +648,9 @@ public Object execute(RequestForm requestForm, BindException errors) throws Exce
         String _newVendorEmail;
         String _newVendorUrl;
         String _newVendorNotes;
+        Double _totalCost;
```
Should this be BigDecimal?

labkey-bpatel replied:
Yes, now changed

labkey-jeckels commented on WNPRC_Purchasing/src/client/components/RequestOrderPanel.tsx:
```diff
@@ -63,7 +65,7 @@ export const RequestOrderPanel: FC<Props> = memo(props => {
     return (
         <>
-        { !isAdmin && canUpdate && (
+        { ((!isAdmin && canUpdate) || (hasRequestId && !canUpdate && !isAdmin)) && (
```
There are a lot of complicated combinations of these checks. Instead of using so many different variants of the check throughout the code, would it be possible to distill them into more understandable booleans? Like canUpdateQuantity, canUpdateAccountAndVendor, etc? If it doesn't improve clarity, feel free to reject this idea.

labkey-bpatel replied:
I agree, it is complicated, I'll go ahead and change it..it is a good idea

labkey-bpatel replied:
Made changes so that it provides lil more clarity on the users and their access to the field.

labkey-jeckels reviewed (approved):
Since I requested changes, I believe that you won't be able to merge this until I approve the updates. Since I'll be out next week, I'm going to mark it as approved. Please ask another developer to review your followup changes based on my suggestions.

labkey-bpatel committed:
Code review changes:
- Use constants instead of hard-coded vals.
- Use getURIString()
- Use getAlias() instead of relying on generated value.
- Other minor changes and cleanup

labkey-bpatel committed:
Code review change - Use Site's default Date format.

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -84,6 +82,7 @@
     private static final Logger _log = LogManager.getLogger(WNPRC_PurchasingController.class);
     private static final DefaultActionResolver _actionResolver = new DefaultActionResolver(WNPRC_PurchasingController.class);
     public static final String NAME = "wnprc_purchasing";
+    public static final Double OVER_FIVE_K = 5000.0;
```
I'd probably call this TOTAL_REQUIRING_ADDITONAL_REVIEW or something. The constant itself isn't over 5K, after all, it's 5K exactly.

And since uses convert it to BigDecimal, why not make the constant a BigDecimal?

labkey-bpatel replied:
true, done

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingController.java:
```diff
@@ -204,7 +202,11 @@ public Object execute(RequestForm requestForm, BindException errors) throws Exce
+
+            //get the lab end user who originated the request
+            List <User> labEndUsers = usersWithInsertPerm.stream().filter(u -> u.getUserId() == emailTemplateForm.getRequester().getUserId()).collect(Collectors.toList());
+            User endUser = labEndUsers.size() == 1 && labEndUsers.get(0) != null ? labEndUsers.get(0) : null;
```
Why do you need to check if labEndUsers.get(0) is null here?

labkey-bpatel replied:
removed

labkey-jeckels reviewed (approved):
A couple more suggestions

labkey-bpatel committed:
Minor bug fixes, code review changes, make display conditions for various users on the data entry page more readable.

labkey-bpatel committed:
Auto test fixes

labkey-bpatel committed:
Update message template format.

labkey-bpatel committed:
Use EmailTemplateService to get the email template obj, instead of creating a new one (otherwise changes made via Customize Email doesn't get reflected).

labkey-bpatel committed:
Handle email notif on line item addition correctly.

labkey-bpatel committed:
optimize imports

labkey-bpatel committed:
Use parameter container instead of getContainer() to get the correct URIString.

labkey-bpatel committed:
Use 'request #' instead of 'request#'

labkey-sweta committed:
WNPRC EMail Notification Automation test

labkey-sweta committed:
WNPRC EMail Notification Automation test

labkey-bpatel committed:
Auto test - add a couple more use cases

labkey-bpatel commented:
@labkey-sweta  Auto test looks good overall. I've added a couple more use cases to the existing tests, and some minor updates since i was already on 21.3 branch.
